src/widgets/reflow.rs

```diff
@@ -72,155 +72,142 @@ where
+                if non_whitespace_previous && is_whitespace
+                    || word_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && self.trim
+                    || whitespace_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && self.trim
+                    || word_width + whitespace_width + symbol_width > self.max_line_width
+                        && current_line.is_empty()
+                        && !self.trim
+                {
+                    if !current_line.is_empty() || !self.trim {
+                        current_line.extend(pending_whitespace.drain(..));
+                        current_line_width += whitespace_width;
                     }
 
-                    // Append remaining text parts
-                    if !unfinished_word.is_empty() || !unfinished_whitespaces.is_empty() {
-                        if current_line.is_empty() && unfinished_word.is_empty() {
-                            wrapped_lines.push(vec![]);
-                        } else if !self.trim || !current_line.is_empty() {
-                            current_line.extend(unfinished_whitespaces.into_iter());
-                        } else {
-                            // TODO: explain why this else branch is ok.
-                            // See clippy::else_if_without_else
+                    current_line.append(&mut pending_word);
+                    current_line_width += word_width;
+
+                    pending_whitespace.clear();
+                    whitespace_width = 0;
+                    word_width = 0;
+                }
+
+                // add finished wrapped line to remaining lines
+                if current_line_width >= self.max_line_width
+                    || current_line_width + whitespace_width + word_width >= self.max_line_width
+                        && symbol_width > 0
+                {
+                    let mut remaining_width =
+                        u16::saturating_sub(self.max_line_width, current_line_width);
+
+                    wrapped_lines.push(std::mem::take(&mut current_line));
+                    current_line_width = 0;
+
+                    // remove whitespace up to the end of line
+                    while let Some(grapheme) = pending_whitespace.front() {
+                        let width = grapheme.symbol.width() as u16;
+
+                        if width > remaining_width {
+                            break;
                         }
-                        current_line.append(&mut unfinished_word);
-                    }
-                    if !current_line.is_empty() {
-                        wrapped_lines.push(current_line);
+
+                        whitespace_width -= width;
+                        remaining_width -= width;
+                        pending_whitespace.pop_front();
                     }
```

Comment:
I tried to change this so that it's just calling pop_front() instead of both front and pop_front. I'm not 100% sure why this isn't the equivalent of:
```rust
// remove whitespace up to the end of line
while let Some(grapheme) = pending_whitespace.pop_front() {
    let width = grapheme.symbol.width() as u16;
    whitespace_width -= width;
    if width > remaining_width {
        break;
    }
    remaining_width -= width;
}
```
This doesn't seem to append the first line of spaces in the indent test, but I'm not sure why.

Reply:
I'm not sure either, most probably something about eagerly popping incorrectly triggers the continue statement below, but it was hard enough to magic a better equivalent of the original code.